### PR TITLE
Add vertexing tutorial

### PR DESCRIPTION
**Pull request by baschlag**
This PR adds a vertexing tutorial MD file as well as dedicated files for writing the tutorial code in.

### CODE REVIEW

**Comment by baschlag** [comment]
@paulgessinger I've adapted the code after our discussion and made it clear that ```TutorialAMVFAlgorithm``` is only a skeleton file to be filled in the tutorial that I've linked in the code. This PR now also introduces a "normal" AMVF example (since this was missing anyways) which always gets built and can be used as reference code for the tutorial.

File: docs/howto/setup_and_run_vertexing.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+This tutorial sets up and runs the ACTS Adaptive Multi-Vertex Finder on smeared ACTS Pythia8 truth tracks.
+*Note*: You have to have Pythia8 available on your machine for following this tutorial.
+## Prerequisites
+For setting up and building ACTS, please refer to the general how-to ACTS guide. For this example you will need to enable Pythia8 in the examples by adding
+```bash
+-DACTS_BUILD_EXAMPLES_PYTHIA8=ON
+```
+to your cmake command.
+
+Note: Additionally adding
```
[the remark this answers is not in the document]
Comment (baschlag) [reply docs/howto/setup_and_run_vertexing.md:11]: Are you sure that this is needed? The default build type should already be `Release`. I would rephrase it to say that it is best to build a `Release` built which is the default. If you have set it to something else, you can use the following settings to set it back to it.

[the remark this answers is not in the document]
Comment (baschlag) [reply docs/howto/setup_and_run_vertexing.md:11]: Are you sure about this? Every time I build Acts from scratch without specifying this option, the execution is extremely slow afterwards and only after rebuilding with ```DCMAKE_BUILD_TYPE=Release```, it speeds up significantly. I couldn't find where the default DCMAKE_BUILD_TYPE is specified though. So I'm not sure about this.

[the remark this answers is not in the document]
Comment (msmk0) [reply docs/howto/setup_and_run_vertexing.md:11]: You are right. I checked again and without an explicit build type no additional build flags are set. 

It would be a good idea to set a default build type e.g. Release if none is given. I'll open an issue for tha.

File: Examples/Algorithms/Vertexing/src/TutorialAMVFAlgorithm.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  std::vector<const Acts::BoundParameters*> inputTrackPtrCollection;
+  for (const auto& trk : allTracks) {
+    inputTrackPtrCollection.push_back(&trk);
+  }
+  //* Do not change the code above this line *//
+
+  //////////////////////////////////////////////////////////////////////////////
+  /*****   Note: This is a skeleton file to be filled with tutorial code  *****/
+  /*****   provided in the Acts Docs - Vertexing section under the link:  *****/
+  /* https://acts.readthedocs.io/en/latest/howto/setup_and_run_vertexing.html */
```
[the remark this answers is not in the document]
Comment (baschlag) [reply Examples/Algorithms/Vertexing/src/TutorialAMVFAlgorithm.cpp:57]: I would prefer to keep this file. There were already several people that followed this tutorial including this hands-on section (after our workshop) and told me that this was extremely useful for them. We could keep it in the Examples for now and could think about introducing a dedicated Tutorial section or so for code like this. I will add a reference to the file in the ```docs/...```.

File: Examples/Algorithms/Vertexing/src/FullExampleAMVFAlgorithm.cpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "ACTFW/Vertexing/FullExampleAMVFAlgorithm.hpp"
+
+#include <Acts/Geometry/GeometryContext.hpp>
```
[the remark this answers is not in the document]
Comment (baschlag) [reply Examples/Algorithms/Vertexing/src/FullExampleAMVFAlgorithm.cpp:11]: > The tutorial looks good.
> 
> I am not sure if we should add the empty tutorial example. I think it is sufficient to have the full example and the detailed description of what each part does in the permanent howto. For the tutorial it made sense to have the empty file but for the permanent version I do not think it does.

See comment below.

[the remark this answers is not in the document]
Comment (baschlag) [reply Examples/Algorithms/Vertexing/src/FullExampleAMVFAlgorithm.cpp:11]: > The two `Acts` includes should also use the `#include "Acts/..."` form.

Done.

File: Examples/Algorithms/Vertexing/src/TutorialAMVFAlgorithm.cpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "ACTFW/Vertexing/TutorialAMVFAlgorithm.hpp"
+
+#include <Acts/Geometry/GeometryContext.hpp>
```
[the remark this answers is not in the document]
Comment (baschlag) [reply Examples/Algorithms/Vertexing/src/TutorialAMVFAlgorithm.cpp:11]: Done.

File: Examples/Run/Vertexing/CMakeLists.txt [hunk 5]
```diff
@@ -26,10 +26,20 @@ target_link_libraries(ActsRecVertexWriter PRIVATE ${_common_libraries})
 add_executable(ActsRecVertexReader VertexAndTracksReaderAndFinderExample.cpp)
 target_link_libraries(ActsRecVertexReader PRIVATE ${_common_libraries})
 
+# Tutorial AMVF example
+add_executable(ActsTutorialAMVF TutorialAMVFExample.cpp)
+target_link_libraries(ActsTutorialAMVF PRIVATE ${_common_libraries})
+
+# Full AMVF example
+add_executable(ActsAdaptiveMultiVertexFinderExample FullAMVFExample.cpp)
```
[the remark this answers is not in the document]
Comment (baschlag) [reply Examples/Run/Vertexing/CMakeLists.txt:34]: Depends on what you call a "full" reco chain... it takes Pythia8 particles, applies a smearing and reconstructs vertices from them. But yes, you're right, the other vertexing example do basically the same, so I will rename it.

[the remark this answers is not in the document]
Comment (baschlag) [reply Examples/Run/Vertexing/CMakeLists.txt]: FYI: In order to make it 100% consistent, I've updated the names of the algorithms and files. The previous ```VertexFindingExample``` is using the ```IterativeVertexFinder``` while the newly introduced example uses the AMVF. All executables, algorithms and example files are now consistently named and the tutorial just refers to our ```AdaptiveMultiVertexFinderAlgorithm```.

File: Examples/Algorithms/Vertexing/src/TutorialAMVFAlgorithm.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (msmk0) [reply Examples/Algorithms/Vertexing/src/TutorialAMVFAlgorithm.cpp:55]: Ok

File: docs/howto/setup_and_run_vertexing.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+to your cmake command will significantly speed up the execution time of the vertexing example algorithm below, it will however slow down the compilation process.
+
+## Setting up an ACTS Adaptive Multi-Vertex Finder Algorithm
+
+A template algorithm file with an (almost) empty ```execute()``` method to be filled in the following is provided here:
+```bash
+../Examples/Algorithms/Vertexing/src/ExampleAMVFAlgorithm.cpp
+```
+Open the file in your editor and let's start setting up the AMVF. We will start setting up all necessary components in the ```execute()``` method.
+*Note:* You would normally **not** want to do all the following setup steps in the ```execute()``` method (that is run on every single event), but rather in some kind of initialization method that is executed only once (which is however not available in the lightweight example framework we will be using here).
```
Comment (msmk0) [inline docs/howto/setup_and_run_vertexing.md:24]: Nitpicking: this kind of initialization should be done in the constructor. Saying this functionality is not available is misleading.

File: Examples/Run/Vertexing/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (msmk0) [reply Examples/Run/Vertexing/CMakeLists.txt]: Excellent. That makes a lot of sense.

File: docs/howto/setup_and_run_vertexing.md (hunk 6 above)
Comment (baschlag) [reply docs/howto/setup_and_run_vertexing.md]: Updated.

**Comment by baschlag** [comment]
Thanks @msmk0, I've updated the docs after your last suggestion. Should be ready from my side.

**Comment by msmk0** [comment]
> Thanks @msmk0, I've updated the docs after your last suggestion. Should be ready from my side.

Thanks, I will merge this once the CI builds finish.

**Comment by msmk0** [comment]
@baschlag Could you rebase the PR?

**Comment by baschlag** [comment]
> @baschlag Could you rebase the PR?

Done.
